src/peft/tuners/oft/layer.py

```diff
@@ -14,106 +14,286 @@
 
 import math
 import warnings
-from typing import Any, List, Optional, Set, Tuple
+from typing import Any, List, Optional, Set, Union
 
 import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
-from peft.tuners.lycoris_utils import LycorisLayer, check_adapters_to_merge
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
 
 
-class OFTLayer(nn.Module, LycorisLayer):
+class MultiplicativeDropoutLayer(nn.Module):
+    """
+    Implements the multiplicative dropout layer for OFT.
+    """
+
+    def __init__(self, p=0.0):
+        """
+        Initializes the multiplicative dropout layer.
+
+        Parameters:
+        p (float): The probability of dropping out a block. Defaults to 0.0.
+        """
+        super().__init__()
+        self.p = p
+
+    def forward(self, x):
+        """
+        Applies multiplicative dropout to the input tensor.
+
+        Parameters:
+        x (Tensor): The input tensor of shape (D, H, H), where `D` represents
+                    the number of OFT blocks, and `H` is the size of the square
+                    blocks along the last two dimensions, the block size in OFT.
+        """
+        if self.training:
+            # Ensure the last two dimensions are the same
+            if x.shape[-1] != x.shape[-2]:
+                raise ValueError("The last two dimensions of input should be the same!")
+
+            D, H, _ = x.shape
+
+            # If block share, skip the multiplicative dropout
+            if D == 1:
+                return x
+
+            num_to_replace = int(self.p * D)
+            num_zeros = D - num_to_replace
+            mask = torch.cat([torch.ones(num_to_replace, device=x.device), torch.zeros(num_zeros, device=x.device)])
+            mask = mask[torch.randperm(D)].view(D, 1, 1)
+            eye_matrix = torch.eye(H, device=x.device).repeat(D, 1, 1)
+            x = (1 - mask) * x + mask * eye_matrix
+        return x
+
+
+class OFTLayer(BaseTunerLayer):
+    """
+    Implements the OFT layer.
+    """
+
     # All names of layers that may contain adapter weights
-    adapter_layer_names = ("oft_r",)
+    adapter_layer_names = ("oft_r", "oft_s")
     # other_param_names is defined on parent class
+    other_param_names = ("r", "oft_block_size", "oft_dropout")
 
-    def __init__(self, base_layer: nn.Module):
-        super().__init__()
-        LycorisLayer.__init__(self, base_layer)
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        """
+        Initializes the OFT layer.
+
+        Note, currently only support linear layer and convolutional layer, with further support for other layers to be
+        added soon.
 
+        Parameters:
+        base_layer: the pretrained model layer
+        """
+        self.base_layer = base_layer
         # OFT info
         self.oft_r = nn.ParameterDict({})
+        self.oft_s = nn.ParameterDict({})
```

Comment:
Here we introduce a new parameter `oft_s`. All existing checkpoints don't have this parameter, so that means when a user loads an old checkpoint, this parameter is randomly initialized, which means the checkpoint behaves different. See my longer comment on compatibility.

Reply:
Actually, it depends on the init_weights, if init_weights=True (which should be the default behavior), then oft_s is basically like an identity and should not make changes. If init_weights=False, then it is randomly initialized.

Reply:
True, but when users load checkpoints, they have trained the model, so `oft_s` won't be an identity transform.